exporter_test.go

```diff
@@ -0,0 +1,429 @@
+		userID:     123,
+		accountID:  456,
+		devices:    []mockDevice{}, // No devices
+		totalWatts: 0,
+		hz:         60.0,
+		voltages:   []float32{120.0, 119.5},
+	}
+	
+	collector := exporter.NewCollector(context.Background(), client, 789, time.Second)
+	metrics := collectMetrics(t, collector)
+	
+	// Verify 'up' is 1
+	verifyMetricValue(t, metrics, "sense_monitor_up", 1.0)
+	
+	// Verify scrape_time_seconds exists
+	if _, exists := metrics["sense_scrape_time_seconds"]; !exists {
+		t.Error("Expected sense_scrape_time_seconds metric to be present")
+	}
+	
+	// Verify monitor metrics exist and report appropriate values
+	verifyMetricValue(t, metrics, "sense_monitor_watts", 0.0)
+	verifyMetricValue(t, metrics, "sense_monitor_hz", 60.0)
+	
+	// Should have voltage metrics for each channel
+	voltageMetrics, exists := metrics["sense_monitor_volts"]
+	if !exists {
+		t.Error("Expected sense_monitor_volts metrics to be present")
+	} else if len(voltageMetrics) != 2 {
+		t.Errorf("Expected 2 voltage metrics, got %d", len(voltageMetrics))
+	}
+	
+	// No device metrics since no devices
+	if _, exists := metrics["sense_device_watts"]; exists {
+		t.Error("Expected no sense_device_watts metrics since no devices")
+	}
+}
+
+func TestCollectorWithDevices(t *testing.T) {
+	// Case 4: Service is up and we have mock devices reporting energy usage
+	devices := []mockDevice{
+		{ID: "light1", Name: "Living Room Light", Type: "Light", Make: "Philips", Model: "Hue", Watts: 25.5, Active: true, Online: true},
+		{ID: "fridge1", Name: "Kitchen Fridge", Type: "Refrigerator", Make: "Samsung", Model: "RF28", Watts: 150.0, Active: true, Online: true},
+		{ID: "washer1", Name: "Washing Machine", Type: "Washer", Make: "LG", Model: "WM3500", Watts: 0, Active: false, Online: false},
+	}
+	
+	client := &mockClient{
+		userID:     123,
+		accountID:  456,
+		devices:    devices,
+		totalWatts: 175.5, // Sum of active device watts
+		hz:         59.8,
+		voltages:   []float32{121.2, 120.8},
+	}
+	
+	collector := exporter.NewCollector(context.Background(), client, 789, time.Second)
+	metrics := collectMetrics(t, collector)
+	
+	// Verify 'up' is 1
+	verifyMetricValue(t, metrics, "sense_monitor_up", 1.0)
+	
+	// Verify scrape_time_seconds exists
+	if _, exists := metrics["sense_scrape_time_seconds"]; !exists {
+		t.Error("Expected sense_scrape_time_seconds metric to be present")
+	}
+	
+	// Verify monitor-level metrics
+	verifyMetricValue(t, metrics, "sense_monitor_watts", 175.5)
+	verifyMetricValue(t, metrics, "sense_monitor_hz", 59.8)
+	
+	// Verify voltage metrics for each channel
+	voltageMetrics, exists := metrics["sense_monitor_volts"]
+	if !exists {
+		t.Error("Expected sense_monitor_volts metrics to be present")
+	} else if len(voltageMetrics) != 2 {
+		t.Errorf("Expected 2 voltage metrics, got %d", len(voltageMetrics))
+	}
+	
+	// Verify device watts metrics - should have one for each device
+	deviceWattsMetrics, exists := metrics["sense_device_watts"]
```

Comment:
Let's create a helper that returns the expectedWatts map

Reply:
Sorry, that's not what I meant.  The code from line 396 to 420 builds a deviceWattsByID struct.  This is what I'm proposing refactoring into a helper.  The actual literal expectedWatts can stay.  The intent here is just to simplify this specific test function so that when you read it you understand what the test semantically is doing.  The mechanics of extracting the data out of the metrics should be in a separate function since that's really not relevant to what's being tested.